keps/sig-node/1999-checkpoint-restore/README.md

```diff
@@ -0,0 +1,277 @@
+
+   As well as the runc support to allow restoring into all kinds of
+   namespaces: https://github.com/opencontainers/runc/pull/2583
+
+   The support to allow restoring into all kinds of namespaces is
+   especially important when checkpointing and restoring containers
+   out of and into pods.
+
+The motivation is to get closer to container migration by taking the first
+step and providing support for simple checkpoint and restore. The motivation
+is existing checkpoint/restore/migration support in other container
+environments showing that it is useful and production ready.
+
+### Goals
+
+The goal of this KEP is to introduce *checkpoint* and *restore* to the API.
+
+### Non-Goals
+
+Out of scope of this KEP are high level discussions about how to implement
+container migration. This is only about low level primitives to add
+*checkpoint* and *restore* to the API.
+
+
+## Proposal
+
+```patch
+--- a/api.proto
++++ b/api.proto
+@@ -87,6 +87,11 @@ service RuntimeService {
+     // Once it returns error, new container log file MUST NOT be created.
+     rpc ReopenContainerLog(ReopenContainerLogRequest) returns (ReopenContainerLogResponse) {}
+ 
++    // CheckpointContainer checkpoints a container
++    rpc CheckpointContainer(CheckpointContainerRequest) returns (CheckpointContainerResponse) {}
++    // RestoreContainer restores a container
++    rpc RestoreContainer(RestoreContainerRequest) returns (RestoreContainerResponse) {}
++
+     // ExecSync runs a command in a container synchronously.
+     rpc ExecSync(ExecSyncRequest) returns (ExecSyncResponse) {}
+     // Exec prepares a streaming endpoint to execute a command in the container.
+@@ -1308,3 +1313,23 @@ message ReopenContainerLogRequest {
+ 
+ message ReopenContainerLogResponse{
+ }
++
++message CheckpointContainerRequest {
++    // ID of the container to be checkpointed
++    string container_id = 1;
++}
++
++message CheckpointContainerResponse{
++}
++
++message RestoreContainerRequest {
++    // ID of the container to be restored
++    string container_id = 1;
++    // ID of the sandbox into which the container should be restored
++    // If this is not specified the container will be restored
++    // in the sandbox it was checkpointed from.
++    string pod_sandbox_id = 2;
++}
++
++message RestoreContainerResponse{
++}
+
+```
+
+There are already draft pull requests opened to show a possible implementation
+of this proposal:
+
+* https://github.com/cri-o/cri-o/pull/4199
+* https://github.com/kubernetes-sigs/cri-tools/pull/662
+
+### User Stories (Optional)
+
+As this KEP is explicitly not about container migration the user stories
+are not mentioning container migration based user stories. If this would
+be helpful to understand this KEP better it can be easily added.
+
+#### Story 1
+
+Although containers are supposed to be stateless there are still containers
+which either require some time to startup or which have data cached.
```

Comment:
Although containers are often considered to be stateless, in many cases containerized applications use in-memory data caching to improve performance (e.g., [Elastisearch](https://www.docker.elastic.co/), [Jira](https://hub.docker.com/r/atlassian/jira-software), [Cassandra](https://hub.docker.com/_/cassandra), [Tomcat](https://hub.docker.com/_/tomcat), [memcached](https://hub.docker.com/_/memcached), [redis](https://hub.docker.com/_/redis), etc.)